### DISCUSSION TITLE
How to enable eager loading for AsyncSession.merge

### DISCUSSION THREAD

**Discussion by daniel-stoneuk** [sqlalchemy/sqlalchemy#6953] [Usage Questions] [answered]
I'm currently migrating some ORM code over to `AsyncSession`, and one of the blockers that I've hit is with `Session.merge` call that I was making previously.

My understanding of `sqlalchemy.orm.Session.merge` is that it will query the database by primary key of the object to be merged. It will then update the attributes on the target object and then returns this updated object. This behaviour appears to be consistent in the `AsyncSession` class.

I have searched through the docs, and unlike `.get` there does not seem to be an `options` parameter that can be used in the database query. Therefore, I have not been able to find a way to enable eager loading for relationship attributes in the returned object. 

In the meantime, I have been using the `run_sync` method like so:

```python
customer = await session.merge(customer)
await session.flush()
...
orders = await session.run_sync(
    lambda s: customer.orders
):
```

Another approach that has worked involved a `session.execute(query ... options(selectinload(Customer.orders))` call after the merge, however, this seems inefficient. (at least to look at - under the hood, I assume SQLAlchemy will use the object in the identity map instead of making another query)

Would anyone be able to offer any advice on the best practice here? Thanks :)

Thanks for such a great library and documentation!

**Comment by zzzeek** [comment, marked as answer]
you've understood the problem perfectly and in fact the "options" that are on "session.get()" are a totally new thing; in legacy SQLAlchemy, get() is a function of the Query object and you would have said query(X).options(...).get().   So when we moved get() to Session I also added that options thing.

so longer term, we should add options to merge() also.   now that get() has this argument, this is a straight passthrough, that is, this is a nearly one-liner PR that would need docstring additions and tests but could be an easy "usecase" addition

```diff
diff --git a/lib/sqlalchemy/orm/session.py b/lib/sqlalchemy/orm/session.py
index af803a1b03..1f00d20d27 100644
--- a/lib/sqlalchemy/orm/session.py
+++ b/lib/sqlalchemy/orm/session.py
@@ -2843,7 +2843,7 @@ class Session(_SessionClassMethods):
             load_options=load_options,
         )
 
-    def merge(self, instance, load=True):
+    def merge(self, instance, load=True, options=None):
         """Copy the state of a given instance into a corresponding instance
         within this :class:`.Session`.
 
@@ -2916,6 +2916,7 @@ class Session(_SessionClassMethods):
                 attributes.instance_state(instance),
                 attributes.instance_dict(instance),
                 load=load,
+                options=options,
                 _recursive=_recursive,
                 _resolve_conflict_map=_resolve_conflict_map,
             )
@@ -2927,6 +2928,7 @@ class Session(_SessionClassMethods):
         state,
         state_dict,
         load=True,
+        options=None,
         _recursive=None,
         _resolve_conflict_map=None,
     ):
@@ -2990,7 +2992,7 @@ class Session(_SessionClassMethods):
                 new_instance = True
 
             elif key_is_persistent:
-                merged = self.get(mapper.class_, key[1], identity_token=key[2])
+                merged = self.get(mapper.class_, key[1], identity_token=key[2], options=options)
 
         if merged is None:
             merged = mapper.class_manager.new_instance()
```

in the meantime, you can keep the SQL emitted down to the minimum by doing the "get" first:

```py
customer = await session.get(customer.id, options=[selectinload(Customer.orders)])
customer = await session.merge(customer)
await session.flush()
...
orders = customer.orders
```

still a workaround though.

**Reply by daniel-stoneuk** [reply]
Brilliant! Thank you for such a thorough answer. I can look into creating the PR, if that would be helpful?

With your example workaround, would this still work if the customer has not been added to the database (and session) yet. Where `customer.id` is not initialised, and the `merge` is used for an "upsert" operation?

I've used the Customer model to produce some more context for my use-case:

```python
""" Change address for customer with name """
# In my use-case, customer_name is unique
customer_name = ...
customer_address = ...

customer_id_row = await session.execute(
    select(customer.id).filter_by(name=customer_name)
)

customer_id = customer_id_row.scalar_one_or_none()

customer = Customer(
    id=customer_id,
    name=customer_name,
    address=customer_address
)

# would the .get work here?
customer = await session.merge(customer)

await session.flush()

do_something(customer.orders)
do_something(customer.id)
```

**Reply by zzzeek** [reply]
OK so you need to check if customer has a primary key already.  if not, then you would not call upon get().    we are basically re-implementing merge a bit.

re: PR I will open an issue you can refer towards in the PR, in #6955 .

**Reply by daniel-stoneuk** [reply]
Thought so. Thanks so much!
